src/utils/messaging-outbox/src/services/implementation/outbox_dispatching_impl.rs

```diff
@@ -0,0 +1,102 @@
+
+use super::{OutboxImmediateImpl, OutboxTransactionalImpl};
+use crate::{MessageConsumer, MessageConsumerMeta, MessageConsumptionDurability, Outbox};
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+pub struct OutboxDispatchingImpl {
+    immediate_outbox: Arc<OutboxImmediateImpl>,
+    transactional_outbox: Arc<OutboxTransactionalImpl>,
+    durable_producers: HashSet<String>,
+    best_effort_producers: HashSet<String>,
+}
+
+#[component(pub)]
+impl OutboxDispatchingImpl {
+    #[allow(clippy::needless_pass_by_value)]
+    pub fn new(
+        catalog: Catalog,
+        immediate_outbox: Arc<OutboxImmediateImpl>,
+        transactional_outbox: Arc<OutboxTransactionalImpl>,
+    ) -> Self {
+        let (durable_producers, best_effort_producers) = Self::classify_message_routes(&catalog);
+
+        Self {
+            immediate_outbox,
+            transactional_outbox,
+            durable_producers,
+            best_effort_producers,
+        }
+    }
+
+    fn classify_message_routes(catalog: &Catalog) -> (HashSet<String>, HashSet<String>) {
+        let mut durable_producers = HashSet::new();
+        let mut best_effort_producers = HashSet::new();
+
+        let all_consumer_builders = catalog.builders_for::<dyn MessageConsumer>();
+        for consumer_builder in all_consumer_builders {
+            let all_metadata: Vec<&MessageConsumerMeta> = consumer_builder.metadata_get_all();
+            assert!(
+                all_metadata.len() <= 1,
+                "Multiple consumer metadata records unexpected for {}",
+                consumer_builder.instance_type_name()
+            );
+            for metadata in all_metadata {
+                for producer_name in metadata.feeding_producers {
+                    match metadata.durability {
+                        MessageConsumptionDurability::Durable => {
+                            durable_producers.insert((*producer_name).to_string());
+                        }
+                        MessageConsumptionDurability::BestEffort => {
+                            best_effort_producers.insert((*producer_name).to_string());
+                        }
+                    }
+                }
+            }
+        }
+
+        (durable_producers, best_effort_producers)
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+impl Outbox for OutboxDispatchingImpl {
+    #[tracing::instrument(level = "debug", skip_all, fields(producer_name, content_json))]
+    async fn post_message_as_json(
+        &self,
+        producer_name: &str,
+        content_json: &serde_json::Value,
+    ) -> Result<(), InternalError> {
+        if self.durable_producers.contains(producer_name) {
+            self.transactional_outbox
+                .post_message_as_json(producer_name, content_json)
```

Comment:
Looks like durability is currently a **consumer-side** parameter (determined by `MessageConsumerMeta`, but shouldn't it be **producer-side**?

Message is either a domain event or a push notification, right?

The nature of the message should be determined from the very beginning by the producer, so that all domain events were saved into DB even if no one is consuming them yet.

Whether a message is a domain event or a notification should not depend on a current set of consumers. After all there might be always zero consumers if a publisher is extracted into a standalone service.

Reply:
I don't agree, currently there are messages that get delivered via both channels, i.e. dataset updates. Overall, this is highly debatable until we get rid of most in-memory implementations.

Reply:
I thought we agreed to pursue an architecture where:
- services can choose between "pure event sourcing" and "state + domain events" design under the hood
- no matter what is primary, events or state, the (public) state is reconcilable via events
- external services synchronize their state via having access to a complete history of such events

I don't see how consumer-side durability fits into this picture. If we write a new service - it has to start producing the outbox events from the start - otherwise when consumers for these events appear they will not have any event history to "catch up" through and will have to have some custom state initialization procedure.

I don't understand how to reason about a system where events are "sometimes persistent" depending on who is listening to them.